builder/builder.go

```diff
@@ -26,24 +27,94 @@ type Builder struct {
 	bc         *blockchain.Blockchain
 	vm         vm.VM
 	newHeads   *feed.Feed[*core.Header]
-	log        utils.Logger
+	log        utils.SimpleLogger
+
+	PendingBlock *blockchain.Pending
 }
 
-func New(ownAddr *felt.Felt, bc *blockchain.Blockchain, builderVM vm.VM, log utils.Logger) *Builder {
+func New(ownAddr *felt.Felt, bc *blockchain.Blockchain, builderVM vm.VM, log utils.SimpleLogger) *Builder {
 	return &Builder{
 		ownAddress: *ownAddr,
 
 		bc:       bc,
 		vm:       builderVM,
 		newHeads: feed.New[*core.Header](),
+		log:      log,
 	}
 }
 
 func (b *Builder) Run(ctx context.Context) error {
-	<-ctx.Done()
+	if err := b.InitPendingBlock(); err != nil {
+		return err
+	}
+
+	blockInterval := 5 * time.Second
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-time.After(blockInterval):
+			if err := b.Finalise(); err != nil {
+				return err
+			}
+		}
+	}
+}
+
+// InitPendingBlock initialises a new pending block
+func (b *Builder) InitPendingBlock() error {
+	nextHeight := uint64(0)
+	parentHash := felt.Zero
+	parentState := felt.Zero
+	if head, err := b.bc.HeadsHeader(); err == nil {
```

Comment:
Other than the first time `InitPendingBlock` is called, `head == b.PendingBlock`. What do you think about making this function stateless by passing in the previous block as a param (`nil` if it doesn't exist)? That way we don't need to read from the db unnecessarily.

Reply:
> But things get trickier when builder is restarted on a non-empty DB

I was thinking that we can call `b.bc.HeadsHeader` only on startup.

Reply:
^Is it harder than calling `b.bc.HeadsHeader` on startup and then passing in `b.PendingBlock.Header` thereafter? If not, I think we might as well do it in this PR.

Reply:
I think we can skip the excursively "empty block" version  of the builder and go straight for normal block building. Closing this PR for now.